src/core-services/account/mutations/createAccountGroup.js

```diff
@@ -0,0 +1,74 @@
+import _ from "lodash";
+import Logger from "@reactioncommerce/logger";
+import ReactionError from "@reactioncommerce/reaction-error";
+import getSlug from "@reactioncommerce/api-utils/getSlug.js";
+
+/**
+ * @name group/createAccountGroup
+ * @method
+ * @memberof Group/Methods
+ * @summary Creates a new permission group for a shop
+ * It creates permission group for a given shop with passed in roles
+ * This method also effectively creates a role in reaction authorization service with the same
+ * name as the supplied group if kafka connect mongo has been configured on the mongo db
+ * @param {object} context - The GraphQL context
+ * @param {String} context.shopId - id of the shop the group belongs to
+ * @param {object} input - The input supplied from GraphQL
+ * @param {Object} input.groupData - info about group to create
+ * @param {String} input.groupData.name - name of the group to be created
+ * @param {String} input.groupData.description - Optional description of the group to be created
+ * @param {Array} input.groupData.permissions - permissions to assign to the group being created
+ * @param {Array} input.groupData.members - members of the
+ * @returns {Object} - `object.status` of 200 on success or Error object on failure
+ */
+export default async function createAccountGroup(context, input) {
+  const { group, shopId } = input;
+  let newlyCreatedGroup;
+  const { Groups } = context.collections;
+
+  // we are limiting group method actions to only users within the account managers role
+  await context.validatePermissions("reaction:account", "create", { shopId, legacyRoles: ["admin"] });
+
+  const defaultCustomerGroupForShop = await Groups.findOne({ slug: "customer", shopId }) || {};
+
+  // TODO: Remove when we move away from legacy permission verification
+  const defaultAdminPermissions = (defaultCustomerGroupForShop.permissions || []).concat("dashboard");
+
+  const nowDate = new Date(Date.now());
+  const newGroupData = Object.assign({}, group, {
+    slug: getSlug(group.name),
+    shopId,
+    createdAt: nowDate,
+    updatedAt: nowDate
+  });
+
+  // TODO: Remove when we move away from legacy permission verification
+  if (!newGroupData.permissions) {
+    newGroupData.permissions = [];
+  }
+
+  // TODO: Remove when we move away from legacy permission verification
+  newGroupData.permissions = _.uniq([...newGroupData.permissions, ...defaultAdminPermissions]);
+
+  // ensure one group type per shop
+  const groupExists = await Groups.findOne({ slug: newGroupData.slug, shopId });
```

Comment:
I think this should be an upsert instead of a find/insert. It would be one operation, and there should be nothing wrong with overwriting data in this case. We would probably want to rename this mutation to `upsertAccountGroup`.

Does anyone have a problem with this? @mikemurray @aldeed 
We may want to do something like this across the codebase, not just here, to save DB operations.

Reply:
Let's just leave it for now, but I am a performance nut and would like to revisit this later, vetting other endpoints we have as well.

Reply:
@rosshadden Since the expectation when calling the mutation is that a group would be created, it's helpful to the client to actually throw an error if that group already exists, versus just updating it silently.

But re performance, I did leave a comment about this code on an earlier commit saying that we could use a dual key unique index on slug + shopId, and then catch the Mongo error below. That would eliminate this findOne.

Reply:
left as is

Reply:
@rosshadden ... as requested ... leaving as is